File: packages/raisin-client-py/raisindb/node_types.py
"""NodeType management operations."""

from typing import TYPE_CHECKING, Any, Dict, List

from .protocol import RequestType, RequestContext

if TYPE_CHECKING:
    from .database import Database


class NodeTypes:
    """NodeType management operations."""

    def __init__(self, database: "Database"):
        """
        Initialize NodeTypes manager.

        Args:
            database: Parent Database instance
        """
        self.database = database

    async def create(self, name: str, node_type: Dict[str, Any]) -> Any:
        """
        Create a new NodeType.

        Args:
            name: NodeType name
            node_type: NodeType definition

        Returns:
            Created NodeType
        """
        context = self.database.get_context()
        payload = {"name": name, "node_type": node_type}
        return await self.database.client.request_tracker.send_request(
            RequestType.NODE_TYPE_CREATE, context, payload
        )

    async def get(self, name: str) -> Any:
        """
        Get a NodeType by name.

        Args:
            name: NodeType name

        Returns:
            NodeType definition
        """
        context = self.database.get_context()
        payload = {"name": name}
        return await self.database.client.request_tracker.send_request(
            RequestType.NODE_TYPE_GET, context, payload
        )

    async def list(self, published_only: bool = False) -> List[Any]:
        """
        List all NodeTypes.

        Args:
            published_only: Only return published NodeTypes

        Returns:
            List of NodeTypes
        """
        context = self.database.get_context()
        payload = {"published_only": published_only}
        result = await self.database.client.request_tracker.send_request(
            RequestType.NODE_TYPE_LIST, context, payload
        )
        return result if isinstance(result, list) else []

    async def update(self, name: str, node_type: Dict[str, Any]) -> Any:
        """
        Update a NodeType.

        Args:
            name: NodeType name
            node_type: Updated NodeType definition

        Returns:
            Updated NodeType
        """
        context = self.database.get_context()
        payload = {"name": name, "node_type": node_type}
        return await self.database.client.request_tracker.send_request(
            RequestType.NODE_TYPE_UPDATE, context, payload
        )

    async def delete(self, name: str) -> Any:
        """
        Delete a NodeType.

        Args:
            name: NodeType name

        Returns:
            Deletion result
        """
        context = self.database.get_context()
        payload = {"name": name}
        return await self.database.client.request_tracker.send_request(
            RequestType.NODE_TYPE_DELETE, context, payload
        )

    async def publish(self, name: str) -> Any:
        """
        Publish a NodeType.

        Args:
            name: NodeType name

        Returns:
            Publish result
        """
        context = self.database.get_context()
        payload = {"name": name}
        return await self.database.client.request_tracker.send_request(
            RequestType.NODE_TYPE_PUBLISH, context, payload
        )

    async def unpublish(self, name: str) -> Any:
        """
        Unpublish a NodeType.

        Args:
            name: NodeType name

        Returns:
            Unpublish result
        """
        context = self.database.get_context()
        payload = {"name": name}
        return await self.database.client.request_tracker.send_request(
            RequestType.NODE_TYPE_UNPUBLISH, context, payload
        )

    async def validate(self, node: Dict[str, Any]) -> Any:
        """
        Validate a node against its NodeType.

        Args:
            node: Node data to validate

        Returns:
            Validation result
        """
        context = self.database.get_context()
        payload = {"node": node}
        return await self.database.client.request_tracker.send_request(
            RequestType.NODE_TYPE_VALIDATE, context, payload
        )

    async def get_resolved(self, name: str) -> Any:
        """
        Get a resolved NodeType with inherited properties.

        Args:
            name: NodeType name

        Returns:
            Resolved NodeType definition
        """
        context = self.database.get_context()
        payload = {"name": name}
        return await self.database.client.request_tracker.send_request(
            RequestType.NODE_TYPE_GET_RESOLVED, context, payload
        )

File: packages/raisin-client-java/src/main/java/com/raisindb/client/operations/NodeTypes.java
package com.raisindb.client.operations;

import com.raisindb.client.protocol.RequestContext;
import com.raisindb.client.protocol.RequestType;

import java.util.*;
import java.util.concurrent.CompletableFuture;

/**
 * NodeType management operations.
 */
public class NodeTypes {

    private final Database database;

    public NodeTypes(Database database) {
        this.database = database;
    }

    /**
     * Create a new NodeType.
     *
     * @param name     NodeType name
     * @param nodeType NodeType definition
     * @return CompletableFuture with created NodeType
     */
    public CompletableFuture<Object> create(String name, Map<String, Object> nodeType) {
        RequestContext context = database.getContext(null, null);
        Map<String, Object> payload = new HashMap<>();
        payload.put("name", name);
        payload.put("node_type", nodeType);

        return database.getClient().getRequestTracker()
                .sendRequest(RequestType.NODE_TYPE_CREATE, context, payload);
    }

    /**
     * Get a NodeType by name.
     *
     * @param name NodeType name
     * @return CompletableFuture with NodeType
     */
    public CompletableFuture<Object> get(String name) {
        RequestContext context = database.getContext(null, null);
        Map<String, Object> payload = new HashMap<>();
        payload.put("name", name);

        return database.getClient().getRequestTracker()
                .sendRequest(RequestType.NODE_TYPE_GET, context, payload);
    }

    /**
     * List all NodeTypes.
     *
     * @param publishedOnly If true, list only published NodeTypes
     * @return CompletableFuture with list of NodeTypes
     */
    public CompletableFuture<List<Object>> list(boolean publishedOnly) {
        RequestContext context = database.getContext(null, null);
        Map<String, Object> payload = new HashMap<>();
        payload.put("published_only", publishedOnly);

        return database.getClient().getRequestTracker()
                .sendRequest(RequestType.NODE_TYPE_LIST, context, payload)
                .thenApply(result -> {
                    if (result instanceof List) {
                        return (List<Object>) result;
                    }
                    return Collections.emptyList();
                });
    }

    /**
     * List all NodeTypes (published and unpublished).
     *
     * @return CompletableFuture with list of NodeTypes
     */
    public CompletableFuture<List<Object>> list() {
        return list(false);
    }

    /**
     * Update a NodeType.
     *
     * @param name     NodeType name
     * @param nodeType Updated NodeType definition
     * @return CompletableFuture with updated NodeType
     */
    public CompletableFuture<Object> update(String name, Map<String, Object> nodeType) {
        RequestContext context = database.getContext(null, null);
        Map<String, Object> payload = new HashMap<>();
        payload.put("name", name);
        payload.put("node_type", nodeType);

        return database.getClient().getRequestTracker()
                .sendRequest(RequestType.NODE_TYPE_UPDATE, context, payload);
    }

    /**
     * Delete a NodeType.
     *
     * @param name NodeType name
     * @return CompletableFuture with result
     */
    public CompletableFuture<Object> delete(String name) {
        RequestContext context = database.getContext(null, null);
        Map<String, Object> payload = new HashMap<>();
        payload.put("name", name);

        return database.getClient().getRequestTracker()
                .sendRequest(RequestType.NODE_TYPE_DELETE, context, payload);
    }

    /**
     * Publish a NodeType.
     *
     * @param name NodeType name
     * @return CompletableFuture with result
     */
    public CompletableFuture<Object> publish(String name) {
        RequestContext context = database.getContext(null, null);
        Map<String, Object> payload = new HashMap<>();
        payload.put("name", name);

        return database.getClient().getRequestTracker()
                .sendRequest(RequestType.NODE_TYPE_PUBLISH, context, payload);
    }

    /**
     * Unpublish a NodeType.
     *
     * @param name NodeType name
     * @return CompletableFuture with result
     */
    public CompletableFuture<Object> unpublish(String name) {
        RequestContext context = database.getContext(null, null);
        Map<String, Object> payload = new HashMap<>();
        payload.put("name", name);

        return database.getClient().getRequestTracker()
                .sendRequest(RequestType.NODE_TYPE_UNPUBLISH, context, payload);
    }

    /**
     * Validate a node against its NodeType.
     *
     * @param node Node to validate
     * @return CompletableFuture with validation result
     */
    public CompletableFuture<Object> validate(Map<String, Object> node) {
        RequestContext context = database.getContext(null, null);
        Map<String, Object> payload = new HashMap<>();
        payload.put("node", node);

        return database.getClient().getRequestTracker()
                .sendRequest(RequestType.NODE_TYPE_VALIDATE, context, payload);
    }

    /**
     * Get resolved NodeType with full inheritance applied.
     *
     * @param name NodeType name
     * @return CompletableFuture with resolved NodeType
     */
    public CompletableFuture<Object> getResolved(String name) {
        RequestContext context = database.getContext(null, null);
        Map<String, Object> payload = new HashMap<>();
        payload.put("name", name);

        return database.getClient().getRequestTracker()
                .sendRequest(RequestType.NODE_TYPE_GET_RESOLVED, context, payload);
    }
}
